### DISCUSSION TITLE
Question about silverstone device support

### DISCUSSION THREAD

**Discussion by matrix200** [arendst/Tasmota#12551] [Q&A] [answered]
Can this device ES03-WiFi  Silverstone be supported?
It is a remote power on/off atx controller that uses usb for power, senses pc power state using pc speaker and uses ewelink app originally.
Has anyone met this device and could tell whether it could be supported with Tasmota?
https://www.silverstonetek.com/product.php?pid=938&area=en#fgo-top
https://www.youtube.com/watch?v=kDylhetcR2Q&t=154s
The device is relatively new.

**Comment by barbudor** [comment, marked as answer]
If it is using an ESP8266 (or ESP32) there it most likely could be supported
Although someone will have to identify the GPIO

But something like that could be build using a mini-D1 with almost no additional parts...

**Comment by matrix200** [comment]
There are two advantages to using this device.
1. Sensing hardware for determininig whether pc is on or off (useful for reporting status).
This uses speaker pins for that.
2. Antenna jack on pcb (it would be pretty much useless in a closed case otherwise).

**Reply by sfromis** [reply]
The product photo includes what looks like the print "PSF-B" on the metal shield at one end of the board. Based on that, I'd expect it to be a PSF-B85 module including an ESP8285 chip (ESP8266 with 1MB of flash on-chip), which would be exactly what Tasmota can run on.

**Reply by matrix200** [reply]
If thats the case, is there any pinout for this particular module?
One can see 4 solder points on the photo, but those appear to be connected to the inputs/outputs used for controlling the power and reset switch and not for serial console.

**Reply by barbudor** [reply]
https://tasmota.github.io/docs/Pinouts/#psf-b85psf-b01psf-b04

**Reply by matrix200** [reply]
Thanks, I guess all that remains is to figure out GPIO pins used.
And the last question.
How does it "sense" whether pc is on or off?
Is there some kind of voltage sensor there, or is it just checking logical "high" or low on the appropriate GPIO pin?

**Reply by barbudor** [reply]
No definitive idea. You would have to do some reverse engineering on the schematic.
But I would assume that it is either connected to a GPIO or the ADC pin and that you will have to "sense" it from Tasmota
To be able to "sense" a GPIO, you will have to define it as a "Switch" and most probably use `SwitchMode 15` (see https://tasmota.github.io/docs/Buttons-and-Switches/)
Shouldn't be too much complicated.

If you can't find out by looking at the board, or testing with a continuity tester, you can start by following instructions in https://tasmota.github.io/docs/Configuration-Procedure-for-New-Devices/

**Reply by barbudor** [reply]
But seriously, 50€ (on Amazon.fr) for something that you can DIY ?

**Reply by matrix200** [reply]
Thats the thing though, I don't think its that easy.
I am not aware of any ready made modules with external antenna connector onboard.
Pretty much any recommendation on adding external antennas goes something like:
1. Make absolutely sure, you know what you are doing.
2. If you can avoid it, don't do it.
3. Adding external antenna to any wifi device (which wasn't designed for it) is hard, requires deep knowledge and experience.
4. If you do it wrong, not only will it not work, but you will mess up the radio spectrum all around yourself.

**Reply by barbudor** [reply]
Very standard
Search for [Mini D1 pro](https://www.aliexpress.com/wholesale?SearchText=mini+D1+pro)
Some are sold with an antenna and a 10cm cable

To drive the reset switch and the power-switch, it's likely a simple resistor + NPN transistor would do the job for each of them.
For the speaker sensing, use the ADC with a 200K ohms resistor in series (to extend the input range from 0..3.3 to 0..5V) and that should be enough.

I don't have any desktop computer anymore, only laptops. So I can't check myself.

**Reply by sfromis** [reply]
Since is connects to "Internal USB 2.0 9 pin" it is probably getting the power state from one (or even more than one) of the pins there.

**Reply by matrix200** [reply]
No, they explicitly say in the manual, that the power status is derived from the speaker pins.
The 9 pin usb plug is actually supposed to be always on (usb standby power) in order for the device to work.

**Reply by matrix200** [reply]
> Very standard
> Search for [Mini D1 pro](https://www.aliexpress.com/wholesale?SearchText=mini+D1+pro)
> Some are sold with an antenna and a 10cm cable
> 
> To drive the reset switch and the power-switch, it's likely a simple resistor + NPN transistor would do the job for each of them.
> For the speaker sensing, use the ADC with a 200K ohms resistor in series (to extend the input range from 0..3.3 to 0..5V) and that should be enough.
> 
> I don't have any desktop computer anymore, only laptops. So I can't check myself.

Excuse me for my ignorance, but how would I go about it?
I suppose I need to hook up the two pins of the power/reset on motherboard to collector and emitter of the transistor respectively, while the base is connected to GPIO right?
The resistor I guess, will have to be on the wire, before the transistor (between the GPIO and transistor).
I would also need to split the wire going from motherboard pins to case somehow, so I could retain the ability to use power and reset buttons?

For speaker, sensing, do I really have to use an extra resistor?
I mean, I don't really need to know exactly what the voltage is.
Just whether the voltage is present or not?

**Reply by barbudor** [reply]
Indeed, going the DIY way rquires some basic skills with electronics like Arduino style

> I suppose I need to hook up the two pins of the power/reset on motherboard to collector and emitter of the transistor respectively, while the base is connected to GPIO right?

Correct.
GPIO ----VVVVV--- Base of NPN transistor
Emitter to GND and Collector to active pin. You will have to probe the connector on your PC to check which one is the GND and which one is the active pin.

> I would also need to split the wire going from motherboard pins to case somehow, so I could retain the ability to use power and reset buttons?

Yes. Indeed the ES03-Wifi is provided with a "Y" cable for that

> For speaker, sensing, do I really have to use an extra resistor?
> I mean, I don't really need to know exactly what the voltage is.

The purpose of the resistor is essentially to reduce the voltage that will reach the ESP8266 pin which is limited to 1V.
On a mini-D1/NodeMCU, there is already a voltage divider which provide an input range of 0..3.3V
That is to say if 3.3V is presented on the min-D1 ADC input, the pin of the ESP will not be more than the 1V it can accept.
If you provide 5V, then you will break it. 
By adding 200k in series (which adds to the already existing 200K and 100k on the mini-D1 board) you will safely reach 0..5V range.

You can also probe the PC power on a disk-drive/floppy connector. If you use the 5V, then you can also make a voltage divider to present 0...3.3V to a standard GPIO that you can probe in Tasmota as a "Switch"

I could make a full schematic of that if you are interested to go this way.

Up to you, you can go for the ES03-Wifi if you want and we will help you the same way.

**Reply by matrix200** [reply]
I think, I will go the DIY way.
Mainly, because I think, flashing the ES03-WIfi will be more problematic then it seems.
Btw I left the same question on Amazon product page and Silverstone responded that the controller is no ESP8266/ESP32 (to be fair I didn't ask about ESP8285).
Supposing that it is, it seems, there are no exposed solder pads for flashing, which means I would have to hook serial port directly to controller pins, which will be very challenging because its so small.
So I will go ahead and order the controller from:
https://www.aliexpress.com/item/32801063577.html?spm=a2g0o.detail.1000014.11.3594b9fcSDOqRl&gps-id=pcDetailBottomMoreOtherSeller&scm=1007.33416.213724.0&scm_id=1007.33416.213724.0&scm-url=1007.33416.213724.0&pvid=2a1e056a-b5b9-415d-b634-11f531ca4434&_t=gps-id:pcDetailBottomMoreOtherSeller,scm-url:1007.33416.213724.0,pvid:2a1e056a-b5b9-415d-b634-11f531ca4434,tpp_buckets:668%230%23208986%234_668%23888%233325%2316_23416%230%23213724%230_23416%236336%2330486%232_23416%234721%2321967%23590_23416%234722%2321972%239_668%232846%238112%231997_668%235811%2327179%2337_668%232717%237566%23868_668%231000022185%231000066059%230_668%233422%2315392%23710_4452%230%23226710%230_4452%233474%2315675%23105_4452%234862%2324463%2393_4452%233098%239599%23133_4452%235108%2323442%23653_4452%233564%2316062%23144 which is one that has antenna included (also the pins).
Can you recommend the right NPN transistor to buy?
https://www.aliexpress.com/wholesale?catId=0&initiative_id=SB_20210704210708&SearchText=npn+transistor
Regarding the Y cable, I think I will try to find one that is ready made.
The wires inside those cables are very thin, I am not sure its a good idea  to strip it of insulation and splitting.
Here is a diagram I drew.
Does it look about right to you?

![circuit](https://user-images.githubusercontent.com/806758/124431086-9f199500-dd78-11eb-9f58-213681da104c.png)

**Reply by matrix200** [reply]
Regarding the Y cable, I suppose I could use a fan splitter cable used to split pwm headers from the motherboard?

**Reply by barbudor** [reply]
mini-D1 : Yes good model
Note that there is still a resistor to move on the D1 to use the external antenna instead of the built-in:
![image](https://user-images.githubusercontent.com/2996042/124434697-77283280-dd74-11eb-9247-6a39840a1cc0.png)

transistor : good old BC547, BC337 or 2N222 will do the job.
You don't need a bag of 50 or 100. Tis is a bag of 5: https://www.aliexpress.com/item/4000025153202.html
But a bag of 50 Is not really more expensive https://www.aliexpress.com/item/1005002120531596.html (take care to choose BC547B)

Regarding your diagram
- Resistor to the base of the transistor is to high. 330 or 470 ohms should be ok. May be 1K also. Depends on the transistor quality
- You got emitter (pin with the arrow) and collector swapped. Emitter to GND and collector to active pin
- I don't understand your connection for the ADC. The ADC pin should go through a 200K resistor to the active pin of the speaker

Regarding power detection, I think you would be better going for picking up a 5V on a free HDD/Floppy connector with a voltage divider:
```
+5V ----R=18K----+----GPIO
GND ----R=33K----+
```
Assuming that the mini-D1 is also connected to PC GND globally by the USB connector

**Reply by matrix200** [reply]
What do you mean by move?
I should desolder the resistor from the board and solder it back using the two pads that you encircled on your photo?
Regarding the emitter/collector, why is it important?
I mean, I could just reverse the connection on the motherboard, no?

Regarding the ADC.
I looked again at what I did on the diagram, I think I am a bit confused there.
What is the reference point for ADC?
So I wire 200K resistor in series to a wire attached to ADC pin and then plug it into the plus side of the speaker?
Just one to one?
No need to wire some ground pin too?
I could use molex connector that I have on the power supply that I don't use.

Not sure about the ground remark.
I am not 100% sure, the usb will provide power to the D1.
In which I case I will have to resort to using an external usb power adapter.
In that case I suppose I will have to wire ground pin on the hdd/floppy/speaker connector to ground pin on the D1 for reference?

**Comment by barbudor** [comment]
> I should desolder the resistor from the board and solder it back using the two pads that you encircled on your photo?

Correct, turn the resistor 90°

> Regarding the emitter/collector, why is it important?
> I mean, I could just reverse the connection on the motherboard, no?

The current flow across a NPN transistor is from collector (input) to emitter (output).
See diagram below where everything that is below "COMPUTER MOTHERBOARD' is actually on the board. I don't have a connector symbol on this diagram editor but understand that the 2 wires from the push button connects to the motherboard connector.
A pushbutton doesn't have a polarity so it doesn't matter how you connect it to the connector. But a transistor has a polarity..
You will have to connect the COLLECTOR to the right side of the switch. The EMITTER should be connected to the GROUND. The common ground between the mini-D1 and the PC. It doesn't have to come from the switch connector header. Even better to have a single GROUND connection between the mini-D1 and those parts and the PC at the USB connector

![image](https://user-images.githubusercontent.com/2996042/124446282-9462fe00-dd80-11eb-935c-f74b78dbd94a.png)

You should check with a voltmeter which pin on the switch cable is high (5V) and which one is GND.
When you press the button, both pin should be GND (the button pull down the active pin to GND)
Better to do this check before as the diagram would be slightly different
